### ISSUE TITLE
Why always --fp32 precision in schedulers?

### ISSUE THREAD

**Issue by himmetozcan** [huggingface/diffusers#3056]
Can anybody clarify why do we always use `--fp32` precision in schedulers such as [here] ?(https://github.com/huggingface/diffusers/blob/4f48476dd6336f35489378cf38c0852a48f92289/src/diffusers/schedulers/scheduling_pndm.py#L112). 
So when I load the model with `--fp16` precision, like this here below, it doesnt affect the schedulers precision?
```
pipe = StableDiffusionInpaintPipeline.from_pretrained(
                "runwayml/stable-diffusion-inpainting",
                torch_dtype=torch.float16
            ).to(device)
```

And another (silly) question. If this is the case, doesn't the model calculations cast to fp32 automatically somewhere in the code, and so slower inference in pytorch code? Or it should give error? or they are casting schedulers back to users input precision somehwere?

**Comment by patrickvonplaten** [comment]
cc @williamberman I think you recently looked into this

**Comment by williamberman** [comment]
parameters of the noise scheduler are very precision sensitive. they need to be calculated in either float32 or even float64 in other diffusion codebases, regardless of the precision used for the model components.

We don't explicitly cast results anywhere for the parameters of the noising process but the schedulers are just implicitly written so the default torch casting rules work with the potentially different precision outputs of the models and the fixed float32 parameters of the noising process. I haven't looked at the explicit casting rule that causes this but I would assume operations with a tensor and a scalar cast to the floating point type of the tensor.


```py
>>> foo = torch.tensor([1,2,3], dtype=torch.float16)
>>> bar = torch.tensor(1, dtype=torch.float32)
>>> foo * bar
tensor([1., 2., 3.], dtype=torch.float16)
>>> foo + bar
tensor([2., 3., 4.], dtype=torch.float16)
>>> bar = torch.tensor(1, dtype=torch.float16)
>>> foo = torch.tensor([1,2,3], dtype=torch.float32)
>>> foo + bar
tensor([2., 3., 4.])
```
